src/app/clusters/water-heater-management-server/water-heater-management-server.cpp

```diff
@@ -0,0 +1,196 @@
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "water-heater-management-server.h"
+
+#include <app/AttributeAccessInterface.h>
+#include <app/AttributeAccessInterfaceRegistry.h>
+#include <app/ConcreteAttributePath.h>
+#include <app/InteractionModelEngine.h>
+#include <app/util/attribute-storage.h>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::WaterHeaterManagement;
+using namespace chip::app::Clusters::WaterHeaterManagement::Attributes;
+
+using chip::Protocols::InteractionModel::Status;
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace WaterHeaterManagement {
+
+CHIP_ERROR Instance::Init()
+{
+    ReturnErrorOnFailure(InteractionModelEngine::GetInstance()->RegisterCommandHandler(this));
+    VerifyOrReturnError(registerAttributeAccessOverride(this), CHIP_ERROR_INCORRECT_STATE);
+
+    return CHIP_NO_ERROR;
+}
+
+void Instance::Shutdown()
+{
+    InteractionModelEngine::GetInstance()->UnregisterCommandHandler(this);
+    unregisterAttributeAccessOverride(this);
+}
+
+bool Instance::HasFeature(Feature aFeature) const
+{
+    return mFeature.Has(aFeature);
+}
+
+// AttributeAccessInterface
+CHIP_ERROR Instance::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    switch (aPath.mAttributeId)
+    {
+    case HeaterTypes::Id:
+        return aEncoder.Encode(mDelegate.GetHeaterTypes());
+    case HeatDemand::Id:
+        return aEncoder.Encode(mDelegate.GetHeatDemand());
+    case TankVolume::Id:
+        if (!HasFeature(Feature::kEnergyManagement))
+        {
+            return CHIP_IM_GLOBAL_STATUS(UnsupportedAttribute);
+        }
+        return aEncoder.Encode(mDelegate.GetTankVolume());
+    case EstimatedHeatRequired::Id:
+        if (!HasFeature(Feature::kEnergyManagement))
+        {
+            return CHIP_IM_GLOBAL_STATUS(UnsupportedAttribute);
+        }
+        return aEncoder.Encode(mDelegate.GetEstimatedHeatRequired());
+    case TankPercentage::Id:
+        if (!HasFeature(Feature::kTankPercent))
+        {
+            return CHIP_IM_GLOBAL_STATUS(UnsupportedAttribute);
+        }
+        return aEncoder.Encode(mDelegate.GetTankPercentage());
+    case BoostState::Id:
+        return aEncoder.Encode(mDelegate.GetBoostState());
+
+    /* FeatureMap - is held locally */
+    case FeatureMap::Id:
+        return aEncoder.Encode(mFeature);
+    }
+
+    /* Allow all other unhandled attributes to fall through to Ember */
+    return CHIP_NO_ERROR;
+}
+
+// CommandHandlerInterface
+CHIP_ERROR Instance::EnumerateAcceptedCommands(const ConcreteClusterPath & cluster, CommandIdCallback callback, void * context)
```

Comment:
Given that this would be handled by the ZAP config correctly if this were not implemented, worth documenting why this code is here.

Reply:
I was not aware that the ZAP config would correctly handle this. Can I just delete this definition of the method as I suspect we're not doing anything special here?

Reply:
Yes.  You need this method if you for some reason need to provide an AcceptedCommandList that does not match the ZAP config.

Reply:
OK/thanks. I have deleted them.